File: leetcode/java/18_4Sum.java
import java.util.*;

class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
        int n = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();

        int k, s, e;
        for (int i = 0; i < n - 3; i++) {
            if (i > 0) {
                while (i < n && nums[i] == nums[i - 1]) {
                    i++;
                }
            }

            for (int j = i + 1; j < n - 2; j++) {
                if (j > i + 1) {
                    while (j < n && nums[j] == nums[j - 1]) {
                        j++;
                    }
                }

                k = target - nums[i] - nums[j];
                s = j + 1;
                e = n - 1;

                while (s < e) {
                    if (nums[s] + nums[e] == k) {
                        ans.add(List.of(nums[i], nums[j], nums[s], nums[e]));
                        s++; e--;
                        while (s < e && nums[s] == nums[s - 1]) s++;
                        while (s < e && nums[e] == nums[e + 1]) e--;
                    } else if (nums[s] + nums[e] < k) {
                        s++;
                    } else {
                        e--;
                    }
                }
            }
        }
        return ans;
    }
}

File: leetcode/cpp/18_4sum.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> fourSum(vector<int>&, int);
};

vector<vector<int>> Solution::fourSum(vector<int>& nums, int target) {
    int n = nums.size();
    sort(nums.begin(), nums.end());
    vector<vector<int>> quadruplets;
   
    int k, start, end;
    
    for (int i = 0; i < n - 3; i++) {
        if (i > 0) while (i < n && nums[i] == nums[i - 1]) i++;
        for (int j = i + 1; j < n - 2; j++) {
            if (j > 0) while (j < n && nums[j] == nums[j - 1]) j++;
            k = target - nums[i] - nums[j];
            
            start = j + 1;
            end = n - 1;
            
            while (start < end) {
                if (nums[start] + nums[end] == k) {
                    quadruplets.push_back({nums[i], nums[j], nums[start], nums[end]});
                    start++; end--;
                    while (start < n && nums[start] == nums[start - 1]) start++;
                    while (end > j && nums[end] == nums[end + 1]) end--;
                } else if (nums[start] + nums[end] > k) {
                    end--;
                    while (end > j && nums[end] == nums[end + 1]) end--;
                } else {
                    start++;
                    while (start < n && nums[start] == nums[start - 1]) start++;
                }
            }
        }
    }
    return quadruplets;
}

int main() {
    vector<int> nums {1, 1, 0, 0, -1, 0, -2, 2};
    Solution *s;
    for (auto& r: s->fourSum(nums, 0)) {
        for (int i: r) cout << i << '\t';
        cout << '\n';
    }
}

[-1,-5,-5,-3,2,5,0,4]

[-5,-5,-3,-1,0,2,4,5]
